initial metrics service

ramaraochavali opened the pull request:
Signed-off-by: Rama <ramaraochavali@gmail.com>

Metrics Service Implementation

Description:
Implements Metrics Service in Envoy that continuously streams metrics to the gRPC end point configured.

Risk Level: Small-medium
Testing: unit test, integration test

*Docs Changes*:
>[Data Plane PR](https://github.com/envoyproxy/data-plane-api/pull/409)

*Release Notes*:
>Release notes updated

*API Changes*:
>[Data Plane API](https://github.com/envoyproxy/data-plane-api/blob/master/api/metrics_service.proto)

ramaraochavali commented:
@mattklein123 @htuch This is by no means complete. But set up the basic stuff along the lines of access log implementation. There are bunch of TODOs to address which i will do in the coming days. But want to get your input at this high level whether direction makes sense before I go too far.
Can you please quickly take a look from that perspective?

dnoe commented:
@mrice32 might like to take a look too

ramaraochavali commented:
@mattklein123 @htuch did some more work.Still few more TODOs to be addressed and tests need to be added.

mrice32 commented on source/common/stats/grpc_metrics_service_impl.h:
```diff
@@ -0,0 +1,169 @@
+
+private:
+  GrpcMetricsStreamerSharedPtr grpc_metrics_streamer_;
+  envoy::api::v2::StreamMetricsMessage message;
```
Nit: `message_;`

htuch commented on source/common/stats/grpc_metrics_service_impl.h:
```diff
@@ -0,0 +1,169 @@
+    void send(envoy::api::v2::StreamMetricsMessage& message);
+
+    GrpcMetricsServiceClientPtr client_;
+    // TODO(ramachavali): Map is not required as there is only one entry.
```
Yeah, I was wondering what the purpose of this was..

ramaraochavali commented:
@mrice32 @htuch reg: thread local, I looked at the statsd sink implementation and it also uses [TLS](https://github.com/envoyproxy/envoy/blob/master/source/common/stats/statsd.cc#L42). So went ahead and used the similar approach that @mattklein123 used for access log implementation for gRPC streaming. I may have misunderstood or missing some thing here. I thought the stats sink should work similar.

mattklein123 commented:
> Aren't counters and gauges flushed to the sink periodically on the main thread? IIUC histogram values may be flushed from many threads, however. So is this TLS work centered around just giving each thread a stream to export its histogram samples, specifically?

That's exactly right. Counters/gauges are flushed on the main thread, however histograms are emitted directly to the sink from each worker.

In the future, when we optionally support built-in histograms, this may not always be the case. A sink could also decide to add the histograms to a queue for later flush by the main thread if desired, but that would add lock contention, unless the queue is per-thread, with periodic flush to the main thread.

I haven't looked at the review yet, but my recommendation would be to probably follow the pattern of the gRPC access logger as was done here I think. We can potentially consider additional designs at a later point. (Note also that this likely will be the same exact design as what will be used by the gRPC tracing code also).

ramaraochavali commented:
@htuch on histogram, i am actually trying to figure out how to exactly map from the envoy implementation to promotheus proto. I could not find direct mapping..will update by tomorrow.

mattklein123 commented:
> Are we planning on doing histograms here? I see a TODO, not sure what the plan is.

If there are no histogram samples emitted in this PR, we don't need TLS. (I have not yet looked at PR).

> Is the implication of this architecture (and the gRPC access logging) that we have one stream per worker? I'm wondering if we'll potentially run into the same issues that inspired ADS in some deployments, namely that streams might map to distinct connections and each connection might point at a different management server, creating additional work at the server. As it is, there's a need to reconcile across the different streams even if only a single management server.

Yup that is the implication. I can't say whether we will run into the same issues, but I know that this design is fine for Lyft. I think if we need aggregation we can do it as a follow up since it will be more complicated (per-thread buffering with flush to central thread, appropriate locking, etc.).

ramaraochavali commented:
I have added integration test as well. So it should be good to review.
Few comments
1. On histogram, currently it is not possible to map Envoy's histogram implementation to [Metrics Proto Histogram](https://github.com/prometheus/client_model/blob/master/metrics.proto#L55) as we do not have buckets defined here. However, if we want to send the raw histogram values we can use [Untyped](https://github.com/prometheus/client_model/blob/master/metrics.proto#L51) metric and send them. Later when Envoy fully supports native histogram implementation this can be changed to use Histogram proto type. Otherwise we can leave histograms for now. Let me know what is the best option here.
2. Regarding the implication that @htuch brought, I think I agree with @mattklein123 that for use cases like metrics, accesslogs etc aggregation may not be required. Even if this stream go to different managed server, finally the managed server is going to push these metrics to backends - it may not matter which managed server does.
3. Even if we leave histograms in this PR, I prefer to keep the TLS implementation as is because when Envoy supports histograms natively, we can just send them much effort., @mattklein123 let me know what you think about this.

Only thing pending from this PR is to update the latest data-plane-api and use the new `GrpcService` that need to be coordinated.

mattklein123 commented on source/common/stats/grpc_metrics_service_impl.h:
```diff
@@ -0,0 +1,167 @@
+#pragma once
```
Side note: Given how much of this code was mostly copied from my access log implementation, I feel like there could be a base class. All of this code is basically going to get copied again for the trace service implementation also. I would just add a TODO somewhere to look into converging this code and the access log code into a common base class.

htuch replied:
+1

mattklein123 commented on source/common/stats/grpc_metrics_service_impl.h:
```diff
@@ -0,0 +1,167 @@
+private:
+  /**
+   * Shared state that is owned by the per-thread streamers. This allows the
+   * main streamer/TLS
```
nit: reflow comment

mattklein123 commented on source/common/stats/grpc_metrics_service_impl.h:
```diff
@@ -0,0 +1,167 @@
+class MetricsServiceSink : public Sink {
+public:
+  // MetricsService::Sink
+  MetricsServiceSink(GrpcMetricsStreamerSharedPtr grpc_metrics_streamer);
```
nit: `const GrpcMetricsStreamerSharedPtr&`

mattklein123 commented on source/common/stats/grpc_metrics_service_impl.h:
```diff
@@ -0,0 +1,167 @@
+    gauage_metric->set_value(value);
+  }
+
+  void endFlush() override { grpc_metrics_streamer_->send(message_); }
```
After the first flush, you should clear identifier info for perf reasons. Since you are reusing the same message as a member that won't happen by default (which is different from how the access log code works).

mattklein123 commented on test/integration/metrics_service_integration_test.cc:
```diff
@@ -0,0 +1,97 @@
+                 metrics_service_request_->headers().Path()->value().c_str());
+    EXPECT_STREQ("application/grpc",
+                 metrics_service_request_->headers().ContentType()->value().c_str());
+    std::cout << "waitForMetricsRequest"
```
del

mattklein123 commented on source/server/config/stats/metrics_service.cc:
```diff
@@ -0,0 +1,73 @@
+};
+
+// Singleton registration via macro defined in envoy/singleton/manager.h
+SINGLETON_MANAGER_REGISTRATION(grpc_metrics_streamer);
```
In the case of the stat sink, you can kill all the singleton stuff. There is only a single sink in the server instantiated.

mattklein123 commented on test/integration/metrics_service_integration_test.cc:
```diff
@@ -0,0 +1,97 @@
+  metrics_service_request_->startGrpcStream();
+  envoy::api::v2::StreamMetricsResponse response_msg;
+  metrics_service_request_->sendGrpcMessage(response_msg);
+  metrics_service_request_->finishGrpcStream(Grpc::Status::Ok);
```
please see https://github.com/envoyproxy/envoy/pull/2333. You will need a similar guard here.

mattklein123 commented on test/integration/metrics_service_integration_test.cc:
```diff
@@ -0,0 +1,97 @@
+    metrics_service_request_ = fake_metrics_service_connection_->waitForNewStream(*dispatcher_);
+  }
+
+  void waitForMetricsRequest() {
```
Can we potentially verify some actual metrics here?

ramaraochavali replied:
i can not verify actual metric names - but i can assert if there are metrics. I think that should be fine. LMK if you think otherwise.

mattklein123 replied:
I think you should be able to verify at least a single gauge and counter exists, and maybe even that a failure is greater than zero. Can you dump them and find some good ones?

ramaraochavali replied:
i think we can actually validate the `metrics_service` cluster related metrics as we are adding the cluster in the test test it self. They will be there for sure. So i have added a validation for a `counter` and `gauge` including values. LMK if that makes sense. Sorry could have done this earlier

ramaraochavali commented:
@mattklein123 addressed all your review comments. can you PTAL?

mattklein123 commented on source/server/config/stats/metrics_service.cc:
```diff
@@ -0,0 +1,68 @@
+                                                                cluster_name),
+          server.threadLocal(), server.localInfo());
+
+  return Stats::SinkPtr(new Stats::Metrics::MetricsServiceSink(grpc_metrics_streamer));
```
nit: std::make_unique

mattklein123 commented on source/server/config/stats/metrics_service.cc:
```diff
@@ -0,0 +1,68 @@
+}
+
+ProtobufTypes::MessagePtr MetricsServiceSinkFactory::createEmptyConfigProto() {
+  return std::unique_ptr<envoy::api::v2::MetricsServiceConfig>(
```
nit: std::make_unique

mattklein123 commented on test/integration/metrics_service_integration_test.cc:
```diff
@@ -0,0 +1,96 @@
+      metrics_service_cluster->mutable_http2_protocol_options();
+
+      auto* metrics_sink = bootstrap.add_stats_sinks();
+      // metrics_sink->MergeFrom(bootstrap.stat_sinks()[0]);
```
?

ramaraochavali replied:
not required. removed.

ramaraochavali commented:
@mattklein123 addressed all the comments. reg actual metrics validation, i think we can actually validate the `metrics_service` cluster related metrics as we are adding the cluster in the test test it self. They will be there for sure. So i have added a validation for a `counter` and `gauge` including values. LMK if that makes sense.

mattklein123 commented on test/integration/metrics_service_integration_test.cc:
```diff
@@ -59,6 +58,34 @@ class MetricsServiceIntegrationTest : public HttpIntegrationTest,
+    std::string known_counter("cluster.metrics_service.membership_change");
+    std::string known_gauge("cluster.metrics_service.membership_total");
+    int metrics_size = envoy_metrics.size();
+    for (int i = 0; i < metrics_size; i++) {
```
nit: You should be able to use a c++11 for loop here and just iterate through the `envoy_metrics()` metrics.

mattklein123 commented on test/integration/metrics_service_integration_test.cc:
```diff
@@ -59,6 +58,34 @@ class MetricsServiceIntegrationTest : public HttpIntegrationTest,
+    int metrics_size = envoy_metrics.size();
+    for (int i = 0; i < metrics_size; i++) {
+      ::io::prometheus::client::MetricFamily metrics_family = envoy_metrics.Get(i);
+      if (known_counter.compare(metrics_family.name()) == 0 &&
```
nit: just do `metrics_family.name() == "cluster.metrics_service.membership_change"`. same below

htuch commented on test/integration/metrics_service_integration_test.cc:
```diff
@@ -0,0 +1,120 @@
+      if (metrics_family.name().compare("cluster.metrics_service.membership_change") == 0 &&
+          metrics_family.metric(0).has_counter()) {
+        known_counter_exists = true;
+        known_counter_value = metrics_family.metric(0).counter().value();
```
I would move the `EXPECT_EQ` to here.

ramaraochavali replied:
moved

htuch commented on test/integration/metrics_service_integration_test.cc:
```diff
@@ -0,0 +1,120 @@
+    bool known_gauge_exists = false;
+    int known_counter_value = -1;
+    int known_gauge_value = -1;
+    for (::io::prometheus::client::MetricFamily metrics_family : envoy_metrics) {
```
It might be slightly cleaner to convert this into a `std::unordered_map` or the like and then write the match logic.

ramaraochavali replied:
i thought about this earlier, but felt building another container may not required here. So did not do. let me know if you feel strongly about it. i can make that change.

ramaraochavali commented on source/common/stats/grpc_metrics_service_impl.cc:
```diff
@@ -0,0 +1,71 @@
+                                                 ThreadLocal::SlotAllocator& tls,
+                                                 const LocalInfo::LocalInfo& local_info)
+    : tls_slot_(tls.allocateSlot()) {
+
```
fixed

ramaraochavali commented:
@htuch addressed all your comments expect using the unorderded map transform in test. let me know if you feel strongly about it. otherwise it is good to go.

ramaraochavali commented:
@htuch can you PTAL when you get time and let me know if any thing else needs to be done on this?

ramaraochavali commented:
@htuch awesome. thanks much..
